serokell-style.md

```diff
@@ -521,7 +571,12 @@ foo = do
 
 ### ApplicativeDo
 
-TODO: use everywhere if possible?
+Don't use `-XApplicativeDo` blindly everywhere. Only when necessary.
+
+You _must_ use `-XApplicativeDo` when you're writing CLI-arguments parser with
```

Comment:
"should" is better.

Reply:
@ChShersh any comments?

Reply:
Well, I think, the better resolution is:

You _must_ use `-XApplicativeDo` if your data type has fields of same type.
You _can_ use `-XApplicativeDo` even if every field of your data type has different arguments (because Applicative style may be shorter).

Reply:
I don't like "must" because it implies i also _must_ reject PRs that don't use extension in parsing. But this is stupid, because it's in fact hard to make a mistake if you're writing optparse parser. Also this kind of mistake is easy to spot. I wouldn't force "must" here. Let's do "should".

Reply:
> I don't like "must" because it implies i also must reject PRs that don't use extension in parsing

Well, I always thought that you _should_ reject PRs which don't follow style-guide, not _must_.

> Also this kind of mistake is easy to spot.

Noticing this error requires thorough code review. If this bug wasn't spot then fixing this bug requires some (probably not so small) amount of time which is bad. While with `-XApplicativeDo` we can completely avoid such errors (or least make it much easier during code review to spot such errors). Also, when arguments are written using applicative style, it requires slightly more time to see correspondence between CLI flag and name of the field inside data type (which is less convenient). So I claim that there're actual benefits in preferring `-XApplicativeDo` over applicative style of creating parser. And, I think that consistent style of code is better.

> Anyway, this is a style document, while the change you suggest is more about "which libraries to use". It's slightly irrelevant.

We already had library-specific requirements before this PR. Also, almost every program we're writing uses CLI. And the only library which is used (and should be used) is `optparse-applicative`. It's not like _if you use `conduit` (not `pipes`) then you should write code like this_. So this library-specific requirement automatically becomes requirement for writing special part of program architecture. I don't know whether such document should exist, like, suggestions for code style when you're using specific library. Maybe style-guide should know nothing about libraries. But we're using such libraries on daily basis. And I think that should have suggestions for how to use libraries. Also, I think it's actually good when usage of some feature/extension/syntax is based on actual use case.

Reply:
>Well, I always thought that you should reject PRs which don't follow style-guide, not must.

In case of PRs "must" is "definitely fix, unless there's a release tomorrow" and "should" is "please fix it, but whatever, here's my approve". 

I still think we should have "should" here. Sometimes it's hard to prove things to people, so let them decide which extension to use. It's the same as with LambdaCase. It's stupid not to use it sometimes, but whatever, it shouldn't be a requirement.

Reply:
Let's hear opinions from other people. If they think that here must me `should` instead of `must` then no problems 👌

Reply:
I am definelly for `should`

Reply:
I'm for "must". The strictness of the requirement isn't a problem by itself, since we do have some policies we enforce (such as spaces vs tabs). Furthermore, I don't think I have seen *any* PRs which don't use `do` with `optparse-applicative`, and if someone chose to do so in the future, it would probably mean that they aren't familiar with the conventions at all, and hence didn't have a strong opinion on the style.